[42기 이한재] Add : 장바구니 기능 구현.

jayhanjaelee opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 데이터베이스 작업
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 - 해당 브랜치(PR)에서 구현하고자 하는 하나의 목표 작성
- email, password를 이용한 로그인 기능 구현 (예시)

<br />

## :: 구현 사항 설명 - 해당 브랜치(PR)에서 작업한 내용 작성

다음 순서대로 로직을 구현하였습니다.

1. UUID 생성
2. transaction start
3. orders 테이블에 데이터 추가
4. orderItems 테이블에 데이터 추가
5. 장바구니에서 데이터 삭제
6. 유저에서 포인트 차감
7. transaction 종료

<br />

## ::  Client Request Body 예시
```json
{
    "cart": [
        {
            "id": 6,
            "userId": 1,
            "quantity": 1,
            "productId": 1,
            "productName": "나이키 에어포스 1 '07",
            "productPrice": 139000,
            "productStyleCode": "CW2288-111",
            "size": "220",
            "stock": 1
        },
        {
            "id": 7,
            "userId": 1,
            "quantity": 1,
            "productId": 2,
            "productName": "에어 조던 1 로우 G",
            "productPrice": 139000,
            "productStyleCode": "CW2288-112",
            "size": "220",
            "stock": 1
        },
        {
            "id": 8,
            "userId": 1,
            "quantity": 1,
            "productId": 3,
            "productName": "나이키 스트릭플라이",
            "productPrice": 139000,
            "productStyleCode": "CW2288-114",
            "size": "220",
            "stock": 1
        },
        {
            "id": 9,
            "userId": 1,
            "quantity": 1,
            "productId": 4,
            "productName": "나이키 에어 덩크 점보",
            "productPrice": 139000,
            "productStyleCode": "CW2288-113",
            "size": "220",
            "stock": 1
        },
        {
            "id": 10,
            "userId": 1,
            "quantity": 1,
            "productId": 5,
            "productName": "나이키 에어 포스 1 '07 LV8",
            "productPrice": 139000,
            "productStyleCode": "CW2288-116",
            "size": "220",
            "stock": 1
        }
    ],
    "shippingAddress": {
        "firstMame": "Hanjae",
        "lastName": "Lee",
        "primary": "충북 청주시 흥덕구 사창동 506",
        "secondary": "201호",
        "province": "충청북도",
        "city": "청주시",
        "zone": "사창동",
        "postalCode": "28643",
        "phoneNumber": "010-9976-9307",
        "email": "jayhanjaelee@gmail.com",    
        "isDefault": false,
        "user_id": 6
    }
}
```

<br />

## :: 기타 질문 및 특이 사항
- 주문하기 과정이 종료되었을 때 response 로 getOrder 함수의 리턴값을 주려고 합니다.
- 그런데 다음과 같은 에러가 나와 문제 해결중에 있습니다.

`TypeError: queryRunner.query is not a function`

- Client Request body 를 위에 코드 블락에 있는 내용으로 가정하여 구현하려고 하는데 수정해야할 부분있을까요?

jayhanjaelee committed:
[42기 이한재] Add : postAddress API 작업중.

jayhanjaelee committed:
Merge branch 'modify/db-schema' into feature/order

jayhanjaelee committed:
Merge branch 'modify/db-schema' into feature/order

jayhanjaelee committed:
[42기 이한재] Add : 주소 추가 & 조회 API 구현.

jayhanjaelee committed:
Merge branch 'main' into feature/order

jayhanjaelee committed:
Merge branch 'main' into feature/order

jayhanjaelee committed:
[42기 이한재] Modify : get product from cart 수정.

jayhanjaelee committed:
Merge branch 'main' into feature/order

jayhanjaelee committed:
[42기 이한재] Add : 주문하기 기능 구현중.

jayhanjaelee committed:
Merge branch 'modify/db-schema' into feature/order

jayhanjaelee committed:
[42기 이한재] Add : 주문하기 기능 구현중.

jayhanjaelee committed:
[42기 이한재] Modify : user class camelCase 로 수정.

jayhanjaelee committed:
[42기 이한재] Add : postOrderItesm 기능 구현.

jayhanjaelee committed:
Merge branch 'main' into feature/order

jayhanjaelee committed:
[42기 이한재] Add : 장바구니 기능 구현.

jayhanjaelee committed:
[42기 이한재] Modify : require 시 .js 삭제.

jayhanjaelee committed:
[42기 이한재] Modify : insertCart func 삭제.

kyeongraekim commented on classes/address.js:
```diff
@@ -0,0 +1,23 @@
+// 4. 배송지에서 필수 유저 정보는
+// 성, 이름, 도로명 주소, 도 / 광역시, 시 / 구 / 군, 우편번호, 전화번호, 이메일
+// 이다.
```
- 상용화를 앞둔 코드에서 불필요한 설명이 담긴 코드 주석은 삭제해주세요! 
- 참고로 실무에서 다량의 코드 주석을 달을 때는 /* asdfasdf  */  기법을 많이 활용하곤 합니다!

jayhanjaelee committed:
[42기 이한재] Modify : isEmpty func -> utils 모듈 함수로 수정.

jayhanjaelee committed:
Merge branch 'main' into feature/order

kyeongraekim commented on classes/user.js:
```diff
@@ -1,11 +1,13 @@
+    this.point = null
   }
 
   keyCheck() {
```
이 부분도 다음의 내용처럼 수정한다면, 불필요한 if-else의 로직 블록을 탈피할 수 있어보이네요!

```javascript
  keyCheck() {
    return Boolean(
      this.firstName && 
      this.lastName && 
      this.email && 
      this.password
      ....
    )

kyeongraekim commented on controllers/orderController.js:
```diff
@@ -0,0 +1,34 @@
+  if (isEmpty(shippingAddress)) throw new Error('emptyAddressErr')
+  if (isEmpty(cart)) throw new Error('EMPTY_CART')
+
+  const orderId = await orderService.postOrders(req.user, shippingAddress, cart)
```
- 5번라인에 할당해 놓은 userId를 사용하면 req 객체에 불필요하게 또 접근할 필요가 없어지겠네요!

kyeongraekim commented on controllers/orderController.js:
```diff
@@ -0,0 +1,34 @@
+  const orderId = await orderService.postOrders(req.user, shippingAddress, cart)
+
+  if (!orderId) {
+    return res.status(200).json({ data: 'Order Failed' })
```
- 실패를 의미하는 상태코드로 변경하는 것이 옳아 보입니다!

kyeongraekim commented on classes/address.js:
```diff
@@ -0,0 +1,23 @@
+// 성, 이름, 도로명 주소, 도 / 광역시, 시 / 구 / 군, 우편번호, 전화번호, 이메일
+// 이다.
+
+class Address {
+  constructor(req) {
+    this.firstName = req?.first_name
+    this.lastName = req?.last_name
+    this.primary = req?.primary
+    this.secondary = req?.secondary
+    this.province = req?.province
+    this.city = req?.city
+    this.zone = req?.zone
+    this.postalCode = req?.postal_code
+    this.phoneNumber = req?.phone_number
+    this.email = req?.email
+    this.isDefault = req?.is_default
+    this.userId = req?.user_id
+  }
+
+}
+
```
- 현재 Address라는 클래스 안에 생성자만 존재할 뿐, 클래스라면 마땅히 지녀야할 내부 메소드, 프로퍼티등의 기능 등이 존재 하지 않고 있어, 클래스를 클래스 답게 쓰지 않고 있는 모습처럼 보여지네요! 
- req.body 안에 원하는 키값이 있는지 없는지 유무를 판단하는 별도의 함수 정도가 이 클래스 내용을 대체하면 좋을 것 같습니다!

kyeongraekim commented on middlewares/auth.js:
```diff
@@ -0,0 +1,25 @@
+      throw new Error('invalidAccessTokenErr')
+    }
+
+    const result = await userDao.getUserById(decoded.user_id)
```
- JWT 토큰을 받아올 때 유저의 pk 값을 의미하는 user_id라는 키값을 그대로 제공하는 것보다, 중의적인 id 정도의 워딩으로 바꾸어도 좋을 것 같습니다!
- 만일 user_id를 유지하실 예정이라면 자바스크립트 컨벤션에 맞게 전달할 때 부터 userId로 전달했음이 옳아 보이네요!

kyeongraekim commented on models/orderDao.js:
```diff
@@ -0,0 +1,182 @@
+        order_status_id)
+    VALUES(?, ?, ?, ?, ?, ?);`
+
+    const orderStatusId = 1
```
- orderStatusId는 별도의 enums.js 라는 파일에 enum의 데이터 형태로 나열하면 좋을 것 같습니다.
- 데이터베이스의 테이블의 내용을 별도의 클래스 또는 객체에 담아서 활용하는 형태라고 보시면 됩니다.
- 해당 내용은 object.freeze 메소드를 활용하여 다음의 형태로 선언해주시면 됩니다.


```javascript
const OrderStatusEnum - Object.freeze (
    PENDING: 1,
    IN_ORDER: 2,
   .... 기타 등등 
)

kyeongraekim commented on models/orderDao.js:
```diff
@@ -0,0 +1,182 @@
+
+const getPaymentAmount = (cart) => {
+  let paymentAmount = 0
+  cart.forEach(c => {
+    paymentAmount += c.productPrice
+  })
```
- 웹 개발 서비스 코딩 시 제 3자를 위해 변수는 늘 원어를 사용해주셔야합니다! c 라는 간단한 for, map 함수에 쓰이는 변수명도 cart 등으로 수정해주세요!

kyeongraekim commented on models/orderDao.js:
```diff
@@ -0,0 +1,182 @@
+  UPDATE users SET point = point - ? WHERE id = ?;`
+
+  const { affectedRows } = await queryRunner.query(rawQuery, [paymentAmount, user.id])
+  if (!affectedRows) throw new Error('NOT_UPDATED_USER_POINT')
```
- affectedRow 유무의 로직으로 현재 모든 dao 내 에러 조건이 설정되어 있는데, affected Row 가 꼭 1이 되어야만 옳은 로직이기 때문에 if ! 보다는 if a == 1 이 로직선에서 오류 발생을 최소화시키는 모습처럼 보여지네요!

jayhanjaelee replied:
`if (affectedRow != 1)` 일때 throw Error 를 해주면 된다는 말씀일까요

아니면
```javascript
if (affectedRow == 1) {
  return
} else {
 // 에러처리
}
```
처럼 적용하라는 말씀이실까요?

kyeongraekim replied:
네 전자의 의미였습니다!

jayhanjaelee replied:
확인했습니다!

kyeongraekim commented on models/userDao.js:
```diff
@@ -50,7 +51,68 @@ const getUser = async (email) => {
+  return result
+}
+
+const postAddress = async (address) => {
+  const keys = []
+  const values = []
+
+  for (let [k, v] of Object.entries(address)) {
+    keys.push('`' + (camelToSnakeCase(k)) + '`')
+    values.push(v)
+  }
```
- 이곳 keys, values의 변수명도 실제 담길 데이터가 드러날 수 있는 변수명을 채택해주세요!
- k, v 와 같은 축약어도 사용을 다시 한 번 지양해주세요!

jayhanjaelee replied:
넵

kyeongraekim commented on services/userService.js:
```diff
@@ -33,16 +33,27 @@ const login = async (email, password) => {
 
     const secretKey = process.env.SECRET_KEY
-    const payLoad = { id: userInfo.id }
+    const payLoad = { user_id: userInfo.id }
```
- 이곳이 제가 앞서 언급드린 부분이겠네요!

kyeongraekim reviewed (changes requested):
- 아래 리뷰 내역 참고하여 수정해주세요!~

jayhanjaelee committed:
[42기 이한재] Refactor : 멘토 리뷰 반영.
- order status enum 추가.
- 불명확한 변수명 변경 ex) keys -> addressKeys
- jwtToken payload user_id -> id 로 변경함에 따라 auth.js 에서도  키 값
  수정.

jayhanjaelee committed:
[42기 이한재] Modify : affectedRows 1이 아닐때 에러 출력하도록  조건문 변경.

jayhanjaelee committed:
Merge branch 'main' into feature/order

jayhanjaelee committed:
Merge branch 'feature/get-products' into feature/order

jayhanjaelee committed:
[42기 이한재] Modify : err 메세지 추가.

jayhanjaelee committed:
Merge branch 'feature/get-products' into feature/order

jayhanjaelee committed:
Merge branch 'feature/get-products' into feature/order

kyeongraekim commented on models/orderDao.js:
```diff
@@ -0,0 +1,180 @@
+      (?, ?, ?, ?, ?);`
+    const { affectedRows } = await queryRunner
+      .query(rawQuery, [c.quantity, c.productPrice, orderId, c.productId, orderStatusId])
+    if (!affectedRows) throw new Error('Not Inserted.')
```
- 이곳의 affectedRow도 아래 내용처럼 내가 영향을 줄 row의 수를 고려해서 설정해주세요!
- 전체 모든 affectedRow에 해당되는 수정 사항내역인데 전체적으로 한번 스캐닝 부탁드립니다~

kyeongraekim replied:
- 나중에 sql injection 공격등 내가 의도하지 않는 열을 건드리는 경우가 생기게되면 affectedRow가 많아질 수도 있는 점을 말씀드립니다!

jayhanjaelee replied:
어떻게 수정하시라는 말씀인지 제대로 이해 못했습니다
forEach 문안에 있는 코드인데 정상적으로 Insert 가 되었다면 affectedRows 가 1 이고 아니라면 0일거라고 생각해서 작성한 코드인데
1보다 더 큰 값이 올수도 있음을 고려해서 작성하시라는 말씀이실까요?
만약 그런 경우라면 어떤식으로 수정하는 것이 좋을까요?

kyeongraekim replied:
@jayhanjaelee 


- AffectedRows는 Boolean 값에 대한 0/1 반환이 아닌, 실제 영향을 받은 Row의 수를 모두 셉니다.
- 즉 한재님이 하나의 row만을 affect 하기로 마음먹은 코드를 짜셨다면, affected row 가 1이 아닌 모든 여집합은 에러의 현상이라고 이해하실 수 있겠습니다.
- 다음의 내용을 참고하여서 affectedRows 를 통해서 어떠한 값이 예상값으로 반환되는지 확인해주세요!


```text
The meaning of the affectedRows value depends on the type of SQL statement that was executed. Here are some examples:

For an INSERT statement, the affectedRows value will be the number of rows that were inserted into the table.
For a DELETE statement, the affectedRows value will be the number of rows that were deleted from the table.
For an UPDATE statement, the affectedRows value will be the number of rows that were updated in the table.
For a SELECT statement, the affectedRows value will be the number of rows that were retrieved by the query.

jayhanjaelee committed:
[42기 이한재] Modify : 주문하기 기능 구현중.

jayhanjaelee committed:
[42기 이한재] Modify : affectedRows 조건문 피드백 반영 및 재고처리에 대한 주문하기 기능 추가.

jayhanjaelee committed:
Merge branch 'main' into feature/order

jayhanjaelee committed:
[42기 이한재] Modify : console.error 삭제.

kyeongraekim reviewed (approved):
LGTM!~ 남은 저녁시간에는 프론트와의 통신에 적극적으로 가담하여 E2E 테스팅을 진행해주세요!~
